src/arith/ir_mutator_with_analyzer.cc

```diff
@@ -20,13 +20,41 @@
 /*!
  * \file tvm/arith/ir_mutator_with_analyzer.cc
  */
+#include <vector>
+
 #include "ir_mutator_with_analyzer.h"
 
 #include <tvm/arith/iter_affine_map.h>
 #include <tvm/tir/analysis.h>
 #include <tvm/tir/op.h>
 
 namespace tvm {
+namespace tir {
+
+bool IsPureCondition(const PrimExpr& condition, const Stmt& scope) {
```

Comment:
For any impure conditional, it looks like we'd end up visiting the body three times (`UsesVar`, `WritesVar`, and then the actual mutation).  For a nested conditional, the nested body would be visited several times.